Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp

```diff
@@ -881,10 +881,13 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
     if (!makeGLContextCurrentIfNeeded())
         return;
 
-    FloatRect rect(tileRect);
-    rect.moveBy(phase);
     SkMatrix phaseMatrix;
-    phaseMatrix.setTranslate(rect.x(), rect.y());
+    // Adjust phase based on tile offset and spacing.
+    double phaseOffsetX = phase.x() + tileRect.x() * patternTransform.a();
+    double phaseOffsetY = phase.y() + tileRect.y() * patternTransform.d();
+    phaseOffsetX -= std::trunc(phaseOffsetX / (tileRect.width() * patternTransform.a() + spacing.width())) * (tileRect.width() * patternTransform.a() + spacing.width());
+    phaseOffsetY -= std::trunc(phaseOffsetY / (tileRect.height() * patternTransform.d() + spacing.height())) * (tileRect.height() * patternTransform.d() + spacing.height());
+    phaseMatrix.setTranslate(phaseOffsetX, phaseOffsetY);
     SkMatrix shaderMatrix = SkMatrix::Concat(phaseMatrix, patternTransform);
     auto samplingOptions = toSkSamplingOptions(m_state.imageInterpolationQuality());
     bool needsClip = tileRect.size() != nativeImage.size();
```

Comment:
If clip is never really needed this variable name is confusing. Since it will be used only once, we don't need it all